[c10d][jit] initial torchbind bindings for ProcessGroupNCCL

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42944 [c10d][jit] initial torchbind bindings for ProcessGroupNCCL**

This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

wanchaol committed:
[WIP][c10d][jit] torchbind ProcessGroup

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"




[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"




[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

gmagogsfm commented on torch/lib/c10d/ProcessGroup.cpp:
```diff
@@ -1,9 +1,34 @@
+// TODO: remove pybind and universally use TorchBind when TorchBind
+// could handle std::chrono like pybind
+static auto allToAllOptions =
+  torch::class_<::c10d::AllToAllOptions>("dist_c10d", "AllToAllOption")
```
typo? AlltToAllOption*s*

gmagogsfm commented on torch/lib/c10d/ProcessGroup.cpp:
```diff
@@ -1,9 +1,34 @@
+  torch::class_<::c10d::AllToAllOptions>("dist_c10d", "AllToAllOption")
+    .def(torch::init<>());
+
+static auto processGroup =
```
I moved a bit further after adding AllToAllOptions TorchBind. 

But then hit issue where ProcessGroup doesn't have "alltoall" defined. After adding "alltoall" binding here, TorchBind complains about unable to handle non-const reference:
https://fburl.com/paste/nqoiu8zt

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

pritamdamania87 commented on test/distributed/test_jit_c10d.py:
```diff
@@ -0,0 +1,75 @@
+    def test_jit_process_group(self):
+        @torch.jit.script
+        def test_create_store(filename: str, world_size: int) -> torch.classes.dist_c10d.FileStore:
+            store = torch.classes.dist_c10d.FileStore(filename, world_size)
```
Is this what the user facing API would look like with torchbind? Is there a way to get around this and use `torch.distributed.FileStore`? The reason being that this is kind of an API change for the user if they would like to use the same code under TorchScript.

wanchaol replied:
ultimately we would like to make `torch.distributed.FileStore = torch.classes.dist_c10d.FileStore`, this way the user could just reuse the existing API either in python and in TorchScript. But this currently is not possible due to various TorchBind limitations (e.g. [support `std::chrono:milisecond` as argument](https://github.com/pytorch/pytorch/issues/42948), [install doc via TorchBind](https://github.com/pytorch/pytorch/issues/42951), etc.) So I am just purely testing it here to make sure the TorchBind part works fine, and we will improve the TorchBind to cover those missing features before fully move to TorchBind

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] torchbind ProcessGroup"


Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] initial torchbind ProcessGroup"




Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP][c10d][jit] initial torchbind ProcessGroup"




Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroup"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroup"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroup"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroup"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroup"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroup"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroup"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroup"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroup"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroup"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroup"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroup"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

gmagogsfm reviewed (approved):
Just a few minor comments discussed offline. LGTM.

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroupNCCL"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroupNCCL"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1223,6 +1223,83 @@ that adds a prefix to each key inserted to the store.
+// document this until we finish the migration.
+
+static const auto StoreTorchBind =
+    torch::class_<::c10d::Store>("dist_c10d", "Store");
```
(not requesting for change)

If `dist_c10d` is what users will see when this is done, let's try to come up with a better name for it. Maybe, sth like `torch.classes.distributed` or `torch.classes.distributed.comm`, etc.

wanchaol replied:
yes I agree with that, the naming is bad. That's something torchbind didn't support yet, we will be adding that support soon and I will switch the names over in follow up PRs.

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1223,6 +1223,83 @@ that adds a prefix to each key inserted to the store.
+#ifdef USE_C10D_NCCL
+
+// XXX: Ideally the Options of ProcessGroupNCCL should be
+// binded using `def_readwrite` like in pybind11, but we
```
typo: binded -> bound

mrshenli commented on test/distributed/test_jit_c10d.py:
```diff
@@ -0,0 +1,73 @@
+        self.world_size = 1
+        self.file = tempfile.NamedTemporaryFile(delete=False)
+        self.num_gpus = torch.cuda.device_count()
+        if self.num_gpus < 2:
```
hmm, doesn't seem the test need two GPUs?

mrshenli commented on test/distributed/test_jit_c10d.py:
```diff
@@ -0,0 +1,73 @@
+            work.wait()
+            return work.result()
+
+        run_pg_nccl_alltoall(nccl_pg, output, input)
```
shall we apply `self.assert*` to check the result?

mrshenli reviewed (commented):
qq, what's the benefit of using torchbind compared to pybind?

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroupNCCL"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d][jit] initial torchbind bindings for ProcessGroupNCCL"


This PR introduces initial custom class bindings for ProcessGroup, and
make it work specifically for ProcessGroupNCCL.

We didn't make the API exactly the same with python pybind bindings, as
there are some issues with TorchBind to support some argument types and
we will resolve that separately to make these two stays the same. And
eventually, when we have a good TorchBind coverage, we will switch all
bindings in c10d TorchBind

Differential Revision: [D23228682](https://our.internmc.facebook.com/intern/diff/D23228682)

[ghstack-poisoned]
